### PR TITLE
gh-103092: Adapt _ctypes's PyCStgDict type to heap type

### PR DESCRIPTION
**Pull request by erlend-aasland**
Port PyCStgDict_Type to heap type.

* Issue: gh-103092

### CODE REVIEW

**Comment by erlend-aasland** [comment]
This probably fails because of the dict hacks in `StructUnionType_new`.

**Comment by vstinner** [comment]
What's the status of this PR? Is it ready for review? It's still marked as a draft.

File: Modules/_ctypes/stgdict.c [hunk 1]
```diff
@@ -36,6 +36,18 @@ PyCStgDict_init(StgDictObject *self, PyObject *args, PyObject *kwds)
     return 0;
 }
 
+static int
+PyCStgDict_traverse(StgDictObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(self));
+    Py_VISIT(self->proto);
```
Comment (neonene) [inline Modules/_ctypes/stgdict.c:44]: Maybe an extra visit to the `proto` member? The `CDataType_traverse` function in _ctypes.c has an explicit `Py_VISIT(dict->proto)`.

**Comment by erlend-aasland** [comment]
I don't have the bandwidth to follow up this right now; abandoning this PR.

**Pull request closed without merging**